### PR TITLE
stage2 Air: add struct_field_ptr_index_{0..3}

### PR DESCRIPTION
**Pull request by g-w1**
Since these are very common, it will save memory.

Mentioned at 2:26 in this stream https://vimeo.com/572815502

"I sped up the compiler by some percent or something"

I added the extra test case to make sure that I did not break structs that did not use the optimization.

### CODE REVIEW

**Comment by andrewrk** [comment]
Prior to merging this, I think we should:
 * Actually use the new AIR instructions
 * Measure a perf improvement with a benchmark

**Comment by g-w1** [comment]
Perf:
test.zig: https://paste.rs/gHy
branch: i-wrote
```
❯ $(which time) -v ../zig-out/bin/zig build-exe test.zig -ofmt=c
	Command being timed: "../zig-out/bin/zig build-exe test.zig -ofmt=c"
	User time (seconds): 0.23
	System time (seconds): 0.07
	Percent of CPU this job got: 100%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.30
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 111296
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 59427
	Voluntary context switches: 180
	Involuntary context switches: 5
	Swaps: 0
	File system inputs: 0
	File system outputs: 55944
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```
branch: master
```
❯ $(which time) -v ../zig-out/bin/zig build-exe test.zig -ofmt=c
	Command being timed: "../zig-out/bin/zig build-exe test.zig -ofmt=c"
	User time (seconds): 0.26
	System time (seconds): 0.07
	Percent of CPU this job got: 104%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.32
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 114828
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 64622
	Voluntary context switches: 100
	Involuntary context switches: 16
	Swaps: 0
	File system inputs: 0
	File system outputs: 56928
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```
It is an improvement!

File: src/codegen/c.zig [hunk 1]
```diff
@@ -1649,15 +1655,29 @@ fn airOptionalPayload(o: *Object, inst: Air.Inst.Index) !CValue {
 
 fn airStructFieldPtr(o: *Object, inst: Air.Inst.Index) !CValue {
     if (o.liveness.isUnused(inst))
-        return CValue.none;
+        return @as(CValue, CValue.none);
```
Comment (andrewrk) [inline src/codegen/c.zig:1658]: Is this change needed?

File: src/codegen/c.zig [hunk 2]
```diff
@@ -1649,15 +1655,29 @@ fn airOptionalPayload(o: *Object, inst: Air.Inst.Index) !CValue {
...
 
     const ty_pl = o.air.instructions.items(.data)[inst].ty_pl;
     const extra = o.air.extraData(Air.StructField, ty_pl.payload).data;
-    const writer = o.writer();
     const struct_ptr = try o.resolveInst(extra.struct_operand);
     const struct_ptr_ty = o.air.typeOf(extra.struct_operand);
+    return airStructFieldPtrFinish(o, inst, struct_ptr_ty, struct_ptr, extra.field_index);
```
Comment (andrewrk) [inline src/codegen/c.zig:1664]:
```suggestion
    return structFieldPtr(o, inst, struct_ptr_ty, struct_ptr, extra.field_index);
```

Reserve "air" prefix for the functions that are called directly from that big switch

File: src/Air.zig [hunk 3]
```diff
@@ -264,6 +264,13 @@ pub const Inst = struct {
         /// Given a pointer to a struct and a field index, returns a pointer to the field.
         /// Uses the `ty_pl` field, payload is `StructField`.
         struct_field_ptr,
+        /// Same as above, but for common cases to save extra data memory.
```
Comment (andrewrk) [inline src/Air.zig:267]: Generally doc comments should reference stuff by name rather than saying "above" or "below" in case someone is looking at documentation out of context, or e.g. looking at generated docs of only 1 field.

Likewise they should be duplicated even for similar things, such as each of these new tags.

File: src/codegen.zig [hunk 4]
```diff
@@ -1597,6 +1603,12 @@ fn Function(comptime arch: std.Target.Cpu.Arch) type {
             //return self.finishAir(inst, result, .{ extra.struct_ptr, .none, .none });
         }
 
+        fn airStructFieldPtrIndex(self: *Self, inst: Air.Inst.Index, index: u8) !void {
+            const ty_op = self.air.instructions.items(.data)[inst].ty_op;
+            _ = ty_op;
+            return self.fail("TODO implement codegen struct_field_ptr_index_{d}", .{index});
```
Comment (andrewrk) [inline src/codegen.zig:1609]: Can you wire this up to the other instruction so that they share a code path?

File: src/codegen/wasm.zig [hunk 5]
```diff
@@ -1443,6 +1447,12 @@ pub const Context = struct {
 
         return WValue{ .local = struct_ptr.multi_value.index + @intCast(u32, extra.data.field_index) };
     }
+    fn airStructFieldPtrIndex(self: *Context, inst: Air.Inst.Index, index: u32) InnerError!WValue {
+        const ty_op = self.air.instructions.items(.data)[inst].ty_op;
+        const struct_ptr = self.resolveInst(ty_op.operand);
+
+        return WValue{ .local = struct_ptr.multi_value.index + index };
```
Comment (andrewrk) [inline src/codegen/wasm.zig:1454]: Let's wire these functions together since there is shared logic (even though it is pretty simple)

**Review by andrewrk — CHANGES_REQUESTED** [review]
Thanks for this. Looks almost ready, just a few cleanups.

**Comment by andrewrk** [comment]
>  20 mili-seconds 

The duration is arbitrary since the test case could be arbitrarily long. So what is interesting is the ratio:

```
0.32 / 0.30 == 1.07
```

So you could say that this improves wall clock time in this execution of this test case by a factor of 1.07.

**Comment by marler8997** [comment]
@g-w1 what's the deal with using `$(which time)` instead of just `time`?
